pgxn/neon/walproposer.c

```diff
@@ -162,6 +176,257 @@ WalProposerCreate(WalProposerConfig *config, walproposer_api api)
+ * thus remove this delay.
+ */
+void
+WalProposerUpdateConfig(WalProposer *wp, WalProposerConfig *config)
+{
+	WalProposerConfig *old = wp->config;
+	bool		changed,
+				reconnect;
+
+	if (config == old)
+		return;
+
+	changed = reconnect = false;
+
+	if (strcmp(old->neon_tenant, config->neon_tenant) != 0)
+	{
+		wp_log(FATAL, "Tenant changed from \"%s\" to \"%s\"",
+			   old->neon_tenant, config->neon_tenant);
+	}
+
+	/*
+	 * Maybe in the future we could support changing the timeline of the node
+	 * on replicas, following timeline's branch point. That's complicated,
+	 * though, so don't bother for now.
+	 */
+	if (strcmp(old->neon_timeline, config->neon_timeline) != 0)
+	{
+		wp_log(FATAL, "Timeline changed from \"%s\" to \"%s\"",
+			   old->neon_tenant, config->neon_tenant);
+	}
+
+	/*
+	 * Maybe in the future we could support changing the timeline of the node
+	 * on replicas, following timeline's branch point. That's complicated,
+	 * though, so don't bother for now.
+	 */
+	if (strcmp(old->safekeepers_list, config->safekeepers_list) != 0)
+	{
+		reconnect = true;
+		changed = true;
+		wp_log(LOG, "Changing safekeepers from [%s] to [%s]",
+			   old->safekeepers_list, config->safekeepers_list);
+	}
+
+	if (old->safekeeper_reconnect_timeout != config->safekeeper_reconnect_timeout)
+	{
+		changed = true;
+	}
+
+	if (old->safekeeper_connection_timeout != config->safekeeper_connection_timeout)
+	{
+		changed = true;
+	}
+
+	if (old->wal_segment_size != config->wal_segment_size)
+	{
+		wp_log(FATAL, "Segment size of running cluster changed");
+	}
+	
+	if (old->syncSafekeepers != config->syncSafekeepers)
+	{
+		/* TODO: Check if this is an interesting workable use case or not */
+	}
+
+	if (old->systemId != config->systemId)
+	{
+		wp_log(FATAL, "System ID of running cluster changed");
+	}
+
+	if (old->pgTimeline != config->pgTimeline)
+	{
+		wp_log(LOG, "PostgreSQL timeline of running cluster changed");
+	}
+
+	/* If nothing changed, don't do anything. */
+	if (!changed)
+	{
+		Assert(!reconnect);
+		return;
+	}
+
+	if (!reconnect)
+	{
+		/*
+		 * We don't store computed results that are based on
+		 * safekeeper_connection_timeout or safekeeper_reconnect_timeout,
+		 * so changes to those values don't need special processing.
+		 *
+		 * If that changes we need to add code here that handles that.
+		 */
+		return;
+	}
+
+	ApplyMergeSKConfigs(wp, config);
+}
+/*
+ * Apply the changes to wp->safekeeper[] slots
+ * We merge configurations into existing safekeepers where possible,
+ * and if they don't map 
+ */
+static void
+ApplyMergeSKConfigs(WalProposer *wp, WalProposerConfig *config)
+{
+	int			len_new = 0;
+	int			found = 0;
+	char	   *new_sk_hosts[MAX_SAFEKEEPERS];
+	int16		map_sk_to_new[MAX_SAFEKEEPERS];
+	int16		map_new_to_sk[MAX_SAFEKEEPERS];
+	int16		sk_reconfig_status[MAX_SAFEKEEPERS];
+
+	len_new = split_safekeepers_list(wp, config->safekeepers_list, new_sk_hosts);
+
+	/* For now, we reboot WP whenever the quorum size changes */
```

Comment:
I think that this limitation doesn't really simplify the code below. It can be removed having almost the same (I would even say more natural) two passes: 1) on first pass iterate over existing wp->safekeepers, finding which ones are not present in the new conf. For each of such sk, terminate connection, log that we are dropping it, importantly reset the state and mark the slot as free. 2) On second pass iterate over new list of safekeepers, for each either finding existing slot if sk was in the old set, or assigning new slot if it wasn't.

State reset is important because otherwise for some time we have confusing/dangerous state where connstr is updated, but state not -- e.g. it might say that sk has given vote which is not really true. Also immediate state reset allows to get rid of pendingReconnects and reconfigureState. Per sk logging about adding/removing sk would be also quite nice.